src/msft-recommendation/src/main/scala/org/apache/spark/ml/recommendation/msft/MsftRecommendation.scala

```diff
@@ -0,0 +1,391 @@
+                        blockSize: Int = 4096): Dataset[Seq[(Int, Array[Float])]] = {
+    import factors.sparkSession.implicits._
+    factors.mapPartitions(_.grouped(blockSize))
+  }
+
+  override def copy(extra: ParamMap): MsftRecommendationModel = {
+    val copied = new MsftRecommendationModel(uid, rank, userFactors, itemFactors)
+    copyValues(copied, extra).setParent(parent)
+  }
+
+  private val predict = udf { (featuresA: Seq[Float], featuresB: Seq[Float]) =>
+    if (featuresA != null && featuresB != null) {
+      // TODO(SPARK-19759): try dot-producting on Seqs or another non-converted type for
+      // potential optimization.
+      blas.sdot(rank, featuresA.toArray, 1, featuresB.toArray, 1)
+    } else {
+      Float.NaN
+    }
+  }
+
+  override def transform(dataset: Dataset[_]): DataFrame = {
+    transformSchema(dataset.schema)
+    // create a new column named map(predictionCol) by running the predict UDF.
+    val predictions = dataset
+      .join(userFactors,
+        checkedCast(dataset($(userCol))) === userFactors("id"), "left")
+      .join(itemFactors,
+        checkedCast(dataset($(itemCol))) === itemFactors("id"), "left")
+      .select(dataset("*"),
+        predict(userFactors("features"), itemFactors("features")).as($(predictionCol)))
+    getColdStartStrategy match {
+      case MsftRecommendationModel.Drop =>
+        predictions.na.drop("all", Seq($(predictionCol)))
+      case MsftRecommendationModel.NaN =>
+        predictions
+    }
+  }
+
+  override def transformSchema(schema: StructType): StructType = {
+    // user and item will be cast to Int
+    SchemaUtils.checkNumericType(schema, $(userCol))
+    SchemaUtils.checkNumericType(schema, $(itemCol))
+    SchemaUtils.appendColumn(schema, $(predictionCol), FloatType)
+  }
+
+  override def write: MsftRecommendationModelWriter = new MsftRecommendationModel.MsftRecommendationModelWriter(this)
+}
+
+object MsftRecommendationModel extends MLReadable[MsftRecommendationModel] {
+
+  private val NaN = "nan"
+  private val Drop = "drop"
+
+  override def read: MsftRecommendationModelReader = new MsftRecommendationModelReader
+
+  override def load(path: String): MsftRecommendationModel = super.load(path)
+
+  private[MsftRecommendationModel] class MsftRecommendationModelWriter(instance: MsftRecommendationModel)
+    extends MLWriter {
+    override protected def saveImpl(path: String): Unit = {
+      val extraMetadata = "rank" -> instance.rank
+      DefaultParamsWriter.saveMetadata(instance, path, sc, Some(extraMetadata))
+      val userPath = new Path(path, "userFactors").toString
+      instance.userFactors.write.format("parquet").save(userPath)
+      val itemPath = new Path(path, "itemFactors").toString
+      instance.itemFactors.write.format("parquet").save(itemPath)
+    }
+
+  }
+
+  private[MsftRecommendationModel] class MsftRecommendationModelReader extends MLReader[MsftRecommendationModel] {
+
+    /** Checked against metadata when loading model */
```

Comment:
A lot of this code might be taken care of by ConstructorWritable and ConstructorReadable. These use reflection to generate serialization code so you dont have to mess with writers. The text featurizer has some usage examples.